src/evict/evict_page.c

```diff
@@ -748,28 +764,37 @@ __evict_review_obsolete_time_window(WT_SESSION_IMPL *session, WT_REF *ref)
     else if (__wt_ref_addr_copy(session, ref, &addr))
         WT_TIME_AGGREGATE_COPY(&newest_ta, &addr.ta);
 
-    /*
-     * Mark the page as dirty to allow the page reconciliation to remove all information related to
-     * an obsolete time window if the page has not been fully removed. The pages that are totally
-     * removed are eliminated during the checkpoint cleanup procedure.
-     */
-    if (!WT_TIME_AGGREGATE_HAS_STOP(&newest_ta) && newest_ta.newest_txn != WT_TXN_NONE &&
-      __wt_txn_visible_all(session, newest_ta.newest_txn,
-        WT_MAX(newest_ta.newest_start_durable_ts, newest_ta.newest_stop_durable_ts))) {
-        __wt_verbose(session, WT_VERB_EVICT,
-          "%p in-memory page obsolete time window: time aggregate %s", (void *)ref,
-          __wt_time_aggregate_to_string(&newest_ta, time_string));
+    /* Pages that are totally removed are eliminated during the checkpoint cleanup procedure. */
+    if (WT_TIME_AGGREGATE_HAS_STOP(&newest_ta))
+        return (0);
 
-        WT_RET(__wt_page_modify_init(session, ref->page));
-        __wt_page_modify_set(session, ref->page);
+    /* If there is no transaction or timestamp information available, there is nothing to do. */
+    max_durable_ts = WT_MAX(newest_ta.newest_start_durable_ts, newest_ta.newest_stop_durable_ts);
+    if (newest_ta.newest_txn == WT_TXN_NONE && max_durable_ts == WT_TS_NONE)
+        return (0);
 
-        /*
-         * To prevent the race while incrementing this variable, no atomic functions are required.
-         * More clean pages may become dirty as a result of an outdated value.
-         */
-        S2BT(session)->obsolete_tw_pages++;
-        WT_STAT_CONN_DSRC_INCR(session, cache_eviction_dirty_obsolete_tw);
-    }
+    /* Ensure the time window information has content that is globally visible. */
+    if (!__wt_txn_visible_all(session, newest_ta.newest_txn, max_durable_ts))
+        return (0);
+
+    /* Mark the page dirty to remove any obsolete information during reconciliation. */
+    __wt_verbose(session, WT_VERB_EVICT,
+      "%p in-memory page obsolete time window: time aggregate %s", (void *)ref,
+      __wt_time_aggregate_to_string(&newest_ta, time_string));
+
+    WT_RET(__wt_page_modify_init(session, ref->page));
+    __wt_page_modify_set(session, ref->page);
+
+    /* Save that a new btree has been processed. */
+    if (__wt_atomic_load32(&btree->obsolete_tw_pages) == 0)
+        __wt_atomic_addv32(&conn->heuristic_controls.obsolete_tw_btree_count, 1);
+
+    /*
+     * To prevent the race while incrementing this variable, no atomic functions are required. More
```

Comment:
It is a bit weird the word "no atomic functions". Isn't __wt_atomic_addv32 an atomic function?

Reply:
Thank you for spotting this, please check f8bd218d491091cb64b9094d67689231afb208c7